### ISSUE TITLE
Elementwise equality comparison with zero-length void dtype fails on NumPy 1.21

### ISSUE THREAD

**Issue by shoyer** [numpy/numpy#19305]
### Reproducing code example:

With NumPy 1.21:
```
>>> import numpy as np
>>> x = np.zeros((3,), np.void)
>>> x == x
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
False
```

I _believe_ the error comes from these lines:
https://github.com/numpy/numpy/blob/1581032a16ba8150f0deca6451f142eeba249dfd/numpy/core/src/multiarray/arrayobject.c#L1361-L1375

On NumPy 1.20, the output is correct:
```
array([ True,  True,  True])
```

Or conversely, if we use a non-zero length void dtype, e.g.,
```
>>> x = np.array([b'foo', b'bar'], np.void)
>>> x == x
array([ True,  True])
```

CC @seberg since this smells similar to https://github.com/numpy/numpy/issues/19085

**Comment by seberg** [comment]
Probably should just fix this local to comparisons, unless you found the issue also in another context?  It is not like you can do a whole lot with voids, especially empty ones...

**Comment by shoyer** [comment]
We have only encountered this in comparisons, so that would work for us.

The empty void dtype is convenient for JAX as a logical "placeholder" for arrays with a particular shape that doesn't require allocating memory (the arrays are always zero).

**Comment by seberg** [comment]
Would be nice if you can test gh-19322, just to be sure.

**Comment by hawkinsp** [comment]
@seberg I can confirm that NumPy built at your PR gh-19322 fixes the test failures we saw in the JAX test suite with NumPy 1.21.

Any chance we can get this fix into 1.21.1 if/when that is a thing?

**Comment by seberg** [comment]
Yes, that PR is practically backport only, the whole branch should use different logic in 1.22/2.0
